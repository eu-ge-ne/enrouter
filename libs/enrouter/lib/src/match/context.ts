import { createContext, useContext } from "react";

import type { Match } from "./mod.js";

const MatchContext = createContext<Match>({
  route: {
    path: "",
    test: { keys: [], pattern: new RegExp("") },
    modules: [],
    loaded: false,
    elements: {},
  },

  isFull: false,
  location: "",
  params: {},
});

export const MatchProvider = MatchContext.Provider;

export function useMatch(): Match {
  return useContext(MatchContext);
}
