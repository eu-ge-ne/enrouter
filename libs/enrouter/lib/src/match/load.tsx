import type { Match } from "./mod.js";
import { loadRoutes } from "#lib/route/load.js";

export function load(matches: Match[]): Promise<void> {
  return loadRoutes(matches.map((x) => x.route).filter((x) => x !== undefined));
}
