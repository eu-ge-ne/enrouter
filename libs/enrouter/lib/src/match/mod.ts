import type { Route } from "#lib/route/mod.js";

/**
 * Represents matched `Route`.
 */
export interface Match {
  route?: Route; // 404?

  isFull: boolean;
  location: string;
  params: Record<string, string>;

  fist?: Match;
  next?: Match;
}
