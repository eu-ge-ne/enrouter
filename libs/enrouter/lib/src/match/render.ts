import type { ReactNode } from "react";

import type { Match } from "./mod.js";

export function renderMatches(matches: Match[]): ReactNode {
  return Object.values(matches[0]?.elements?.layout ?? {})[0];
}
