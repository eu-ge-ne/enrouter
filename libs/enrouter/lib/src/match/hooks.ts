import type { Match } from "./mod.js";
import { useMatch } from "./context.js";

export function useRoot(): Match | undefined {
  let match = useMatch()?.last;

  do {
    if (match?.route.elements._root) {
      return match;
    }
    match = match?.prev;
  } while (match);
}

export function usePath(path: string): Match | undefined {
  let match = useMatch()?.first;

  while (match) {
    if (match.route?.path === path) {
      return match;
    }

    match = match.next;
  }
}

export interface UseActiveParams<T> {
  path: string;
  loose?: boolean;
  value: [T, T];
}

export function useActive<T>({ path, loose, value }: UseActiveParams<T>): T {
  const match = usePath(path);

  const isActive = loose ? match !== undefined : Boolean(match?.full);

  return isActive ? value[0] : value[1];
}
