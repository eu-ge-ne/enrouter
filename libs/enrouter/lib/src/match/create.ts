import type { Route } from "#lib/route/mod.js";
import type { Match } from "./mod.js";
import { logger } from "#lib/debug.js";
import { loadRoutes } from "#lib/route/load.js";

const log = logger("match/create");

export interface CreateMatchParams {
  routes: Route;
  location: string;
}

export async function createMatch({
  routes,
  location,
}: CreateMatchParams): Promise<Match | undefined> {
  const matches: Match[] = [];

  recur([routes], location, matches);

  await loadRoutes(matches.map((x) => x.route));

  trim(matches);
  link(matches);

  log("matched: %o", matches);

  return matches[0];
}

function recur(routes: Route[], location: string, matches: Match[]): void {
  const match = matchOneOf(routes, location);
  if (!match) {
    return;
  }

  log(`path: "%s", location: "%s"`, match.route.path, match.location);

  matches.push(match);

  if (match.route.tree) {
    recur(match.route.tree, location, matches);
  }
}

function matchOneOf(routes: Route[], location: string): Match | undefined {
  let matched = routes
    .map((route) => {
      const execs = route.test.pattern.exec(location);
      if (execs) {
        return { route, execs };
      }
    })
    .filter((x) => x !== undefined);

  if (matched.length === 0) {
    return;
  }

  // TODO: improve

  if (matched.length > 1) {
    matched = matched.filter((x) => x.route.test.keys.length === 0);
  }

  if (matched.length !== 1) {
    throw new Error("Found >1 match");
  }

  const { route, execs } = matched[0]!;
  const loc = execs[0] || "/";

  return {
    route,
    isFull: loc === location,
    location: loc,
    params: Object.fromEntries(
      route.test.keys.map((key, i) => [key, execs![i + 1]!] as const),
    ),
  };
}

function trim(matches: Match[]): void {
  const rootIndex = matches.findLastIndex((x) => x.route.elements._root);
  if (rootIndex >= 0) {
    matches.splice(0, rootIndex);
  }

  if (!matches.at(-1)?.isFull) {
    const voidIndex = matches.findLastIndex(
      ({
        route: {
          elements: { _void, __void },
        },
      }) => _void || __void,
    );

    if (voidIndex >= 0) {
      if (matches[voidIndex]?.route.elements.__void) {
        const match = matches[voidIndex];
        matches.splice(0, matches.length);
        matches.push(match);
      } else {
        matches.splice(voidIndex + 1);
      }
    }
  }
}

function link(matches: Match[]): void {
  const first = matches[0];
  const last = matches.at(-1);

  matches.forEach((x, i) => {
    x.first = first;
    x.next = matches[i + 1];
    x.last = last;
  });
}
