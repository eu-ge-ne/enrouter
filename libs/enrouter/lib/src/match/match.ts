import type { Route } from "#lib/route/mod.js";
import type { Match } from "./mod.js";
import { logger } from "#lib/debug.js";
import { load } from "./load.js";

const log = logger("match");

export interface MatchParams {
  routes: Route;
  location: string;
}

export async function match({
  routes,
  location,
}: MatchParams): Promise<Match[]> {
  const matches: Match[] = [];

  recur([routes], location, matches);

  await load(matches);

  if (!matches.at(-1)?.isFull) {
    const i = matches.findLastIndex(
      (x) => Object.values(x.route.elements.notFound ?? {}).length > 0,
    );
    matches.splice(Math.max(1, i + 1));
  }

  log("matched: %o", matches);

  matches.forEach((x, i) => {
    x.first = matches[0];
    x.next = matches[i + 1];
  });

  return matches;
}

function recur(routes: Route[], location: string, matches: Match[]): void {
  const match = matchOneOf(routes, location);
  if (!match) {
    return;
  }

  log(`path: "%s", location: "%s"`, match.route?.path, match.location);

  matches.push(match);

  if (match?.route?.tree) {
    recur(match.route.tree, location, matches);
  }
}

function matchOneOf(routes: Route[], location: string): Match | undefined {
  let matched = routes
    .map((route) => {
      const execs = route.test.pattern.exec(location);
      if (!execs) {
        return;
      }
      return { route, execs };
    })
    .filter((x) => x !== undefined);

  if (matched.length === 0) {
    return;
  }

  // TODO: improve

  if (matched.length > 1) {
    matched = matched.filter((x) => x.route.test.keys.length === 0);
  }

  if (matched.length !== 1) {
    throw new Error("Found >1 match");
  }

  const { route, execs } = matched[0]!;
  const loc = execs[0] || "/";

  return {
    route,
    isFull: loc === location,
    location: loc,
    params: Object.fromEntries(
      route.test.keys.map((key, i) => [key, execs![i + 1]!] as const),
    ),
  };
}
