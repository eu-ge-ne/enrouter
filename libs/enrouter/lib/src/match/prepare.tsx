import type { ReactElement } from "react";

import type { Match } from "./mod.js";
import { loadRoutes } from "#lib/route/load.js";
import { MatchProvider } from "./context.js";

export async function prepareMatches(matches: Match[]) {
  await loadRoutes(matches.map((x) => x.route).filter((x) => x !== undefined));

  for (let i = matches.length - 1; i >= 0; i -= 1) {
    const match = matches[i]!;

    const wrap = (els?: Record<string, ReactElement>) =>
      els
        ? Object.fromEntries(
            Object.entries(els).map(([key, children]) => [
              key,
              <MatchProvider key={key} value={match}>
                {children}
              </MatchProvider>,
            ]),
          )
        : undefined;

    match.elements = {
      layout: wrap(match.route?.elements.layout),
      index: wrap(match.route?.elements.index),
      notFound: wrap(match.route?.elements.notFound),
    };
  }
}
