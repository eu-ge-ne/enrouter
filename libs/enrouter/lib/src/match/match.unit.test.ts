import { describe, test, expect } from "vitest";
import * as regexparam from "regexparam";

import type { Route } from "#lib/route/mod.js";
import { match } from "./match.js";

describe("match", () => {
  describe("match", () => {
    test("0 matches", async () => {
      const routes: Route = {
        path: "/abc",
        test: regexparam.parse("/abc", true),
        modules: [],
        loaded: false,
        elements: {},
      };

      expect(await match({ routes, location: "/x" })).toMatchSnapshot();
    });

    test("1 match", async () => {
      const routes: Route = {
        path: "/",
        test: regexparam.parse("/", true),
        modules: [],
        loaded: false,
        elements: {},
      };

      expect(await match({ routes, location: "/" })).toMatchSnapshot();
    });

    test("2 matches", async () => {
      const routes: Route = {
        path: "/",
        test: regexparam.parse("/", true),
        modules: [],
        loaded: false,
        elements: {},
        tree: [
          {
            path: "/abc",
            test: regexparam.parse("/abc", true),
            modules: [],
            loaded: false,
            elements: {},
          },
        ],
      };

      expect(await match({ routes, location: "/abc" })).toMatchSnapshot();
    });

    test("1 match with params", async () => {
      const routes: Route = {
        path: "/[:id]",
        test: regexparam.parse("/:id", true),
        modules: [],
        loaded: false,
        elements: {},
      };

      expect(await match({ routes, location: "/100" })).toMatchSnapshot();
    });

    test("1 match and 1 match with params", async () => {
      const routes: Route = {
        path: "/",
        test: regexparam.parse("/", true),
        modules: [],
        loaded: false,
        elements: {},
        tree: [
          {
            path: "/abc",
            test: regexparam.parse("/abc", true),
            modules: [],
            loaded: false,
            elements: {},
          },
          {
            path: "/[:id]",
            test: regexparam.parse("/:id", true),
            modules: [],
            loaded: false,
            elements: {},
          },
        ],
      };

      expect(await match({ routes, location: "/abc" })).toMatchSnapshot();
    });
  });
});
