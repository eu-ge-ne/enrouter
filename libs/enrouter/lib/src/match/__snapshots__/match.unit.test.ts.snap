// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`match > match > 0 matches 1`] = `[]`;

exports[`match > match > 1 match 1`] = `
[
  {
    "first": [Circular],
    "isFull": true,
    "location": "/",
    "next": undefined,
    "params": {},
    "route": {
      "elements": {},
      "loaded": false,
      "modules": [],
      "path": "/",
      "test": {
        "keys": [],
        "pattern": /\\^\\(\\?=\\$\\|\\\\/\\)/i,
      },
    },
  },
]
`;

exports[`match > match > 1 match and 1 match with params 1`] = `
[
  {
    "first": [Circular],
    "isFull": false,
    "location": "/",
    "next": {
      "first": [Circular],
      "isFull": true,
      "location": "/abc",
      "next": undefined,
      "params": {},
      "route": {
        "elements": {},
        "loaded": false,
        "modules": [],
        "path": "/abc",
        "test": {
          "keys": [],
          "pattern": /\\^\\\\/abc\\(\\?=\\$\\|\\\\/\\)/i,
        },
      },
    },
    "params": {},
    "route": {
      "elements": {},
      "loaded": false,
      "modules": [],
      "path": "/",
      "test": {
        "keys": [],
        "pattern": /\\^\\(\\?=\\$\\|\\\\/\\)/i,
      },
      "tree": [
        {
          "elements": {},
          "loaded": false,
          "modules": [],
          "path": "/abc",
          "test": {
            "keys": [],
            "pattern": /\\^\\\\/abc\\(\\?=\\$\\|\\\\/\\)/i,
          },
        },
        {
          "elements": {},
          "loaded": false,
          "modules": [],
          "path": "/[:id]",
          "test": {
            "keys": [
              "id",
            ],
            "pattern": /\\^\\\\/\\(\\[\\^/\\]\\+\\?\\)\\(\\?=\\$\\|\\\\/\\)/i,
          },
        },
      ],
    },
  },
  {
    "first": {
      "first": [Circular],
      "isFull": false,
      "location": "/",
      "next": [Circular],
      "params": {},
      "route": {
        "elements": {},
        "loaded": false,
        "modules": [],
        "path": "/",
        "test": {
          "keys": [],
          "pattern": /\\^\\(\\?=\\$\\|\\\\/\\)/i,
        },
        "tree": [
          {
            "elements": {},
            "loaded": false,
            "modules": [],
            "path": "/abc",
            "test": {
              "keys": [],
              "pattern": /\\^\\\\/abc\\(\\?=\\$\\|\\\\/\\)/i,
            },
          },
          {
            "elements": {},
            "loaded": false,
            "modules": [],
            "path": "/[:id]",
            "test": {
              "keys": [
                "id",
              ],
              "pattern": /\\^\\\\/\\(\\[\\^/\\]\\+\\?\\)\\(\\?=\\$\\|\\\\/\\)/i,
            },
          },
        ],
      },
    },
    "isFull": true,
    "location": "/abc",
    "next": undefined,
    "params": {},
    "route": {
      "elements": {},
      "loaded": false,
      "modules": [],
      "path": "/abc",
      "test": {
        "keys": [],
        "pattern": /\\^\\\\/abc\\(\\?=\\$\\|\\\\/\\)/i,
      },
    },
  },
]
`;

exports[`match > match > 1 match with params 1`] = `
[
  {
    "first": [Circular],
    "isFull": true,
    "location": "/100",
    "next": undefined,
    "params": {
      "id": "100",
    },
    "route": {
      "elements": {},
      "loaded": false,
      "modules": [],
      "path": "/[:id]",
      "test": {
        "keys": [
          "id",
        ],
        "pattern": /\\^\\\\/\\(\\[\\^/\\]\\+\\?\\)\\(\\?=\\$\\|\\\\/\\)/i,
      },
    },
  },
]
`;

exports[`match > match > 2 matches 1`] = `
[
  {
    "first": [Circular],
    "isFull": false,
    "location": "/",
    "next": {
      "first": [Circular],
      "isFull": true,
      "location": "/abc",
      "next": undefined,
      "params": {},
      "route": {
        "elements": {},
        "loaded": false,
        "modules": [],
        "path": "/abc",
        "test": {
          "keys": [],
          "pattern": /\\^\\\\/abc\\(\\?=\\$\\|\\\\/\\)/i,
        },
      },
    },
    "params": {},
    "route": {
      "elements": {},
      "loaded": false,
      "modules": [],
      "path": "/",
      "test": {
        "keys": [],
        "pattern": /\\^\\(\\?=\\$\\|\\\\/\\)/i,
      },
      "tree": [
        {
          "elements": {},
          "loaded": false,
          "modules": [],
          "path": "/abc",
          "test": {
            "keys": [],
            "pattern": /\\^\\\\/abc\\(\\?=\\$\\|\\\\/\\)/i,
          },
        },
      ],
    },
  },
  {
    "first": {
      "first": [Circular],
      "isFull": false,
      "location": "/",
      "next": [Circular],
      "params": {},
      "route": {
        "elements": {},
        "loaded": false,
        "modules": [],
        "path": "/",
        "test": {
          "keys": [],
          "pattern": /\\^\\(\\?=\\$\\|\\\\/\\)/i,
        },
        "tree": [
          {
            "elements": {},
            "loaded": false,
            "modules": [],
            "path": "/abc",
            "test": {
              "keys": [],
              "pattern": /\\^\\\\/abc\\(\\?=\\$\\|\\\\/\\)/i,
            },
          },
        ],
      },
    },
    "isFull": true,
    "location": "/abc",
    "next": undefined,
    "params": {},
    "route": {
      "elements": {},
      "loaded": false,
      "modules": [],
      "path": "/abc",
      "test": {
        "keys": [],
        "pattern": /\\^\\\\/abc\\(\\?=\\$\\|\\\\/\\)/i,
      },
    },
  },
]
`;
